experimental/web_dashboard/lib/src/experiments/routing_demo.dart

```diff
@@ -0,0 +1,207 @@
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/scheduler.dart';
+
+import '../widgets/third_party/adaptive_scaffold.dart';
+
+/// The [Key] for the inner [Navigator].
+GlobalKey _navigatorKey = GlobalKey();
+
+void main() {
+  runApp(RoutingDemoApp());
+}
+
+/// A responsive dashboard using AdaptiveScaffold and NavigationRail with named
+/// routes. Routing works as expected using a few tricks.
+class RoutingDemoApp extends StatefulWidget {
+  @override
+  _RoutingDemoAppState createState() => _RoutingDemoAppState();
+}
+
+class _RoutingDemoAppState extends State<RoutingDemoApp> {
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      // Forward the initial route to the app shell.
+      onGenerateInitialRoutes: (initialRoute) {
+        return [
+          PageRouteBuilder(
+            // Use the same route settings in the child navigator.
+            settings: RouteSettings(name: initialRoute),
+            // Send the route to AppShell so it can select the correct item in
+            // the Drawer / NavigationRail / BottomAppBar.
+            pageBuilder: (context, animation, secondaryAnimation) {
+              return AppShell(initialRoute);
+            },
+          ),
+        ];
+      },
+    );
+  }
+}
+
+/// The outer "shell" of the app. Uses an [AdaptiveScaffold] with a [Navigator]
+/// to handle route events and display the correct page in the Drawer /
+/// BottomNavigationBar, or NavigationRail, as well as the correct page in the
+/// body of the Scaffold.
+class AppShell extends StatefulWidget {
+  /// The name of the initial route. Required to set the selected index in the
+  /// sidebar.
+  final String initialRouteName;
+
+  AppShell(this.initialRouteName);
+
+  @override
+  _AppShellState createState() => _AppShellState();
+}
+
+class _AppShellState extends State<AppShell> {
+  int _currentIndex = 0;
+
+  void initState() {
+    super.initState();
+    _currentIndex = _indexForRoute(widget.initialRouteName);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return AdaptiveScaffold(
+      title: Text('Dashboard App'),
+      body: Navigator(
+        key: _navigatorKey,
+        initialRoute: widget.initialRouteName,
+        onGenerateRoute: (settings) {
+          var routeName = settings.name;
+          var pageWidget = _pageForRoute(routeName);
+
+          // Change the _currentIndex to highlight the correct menu item /
+          // BottomNavigationBarItem.
+          //
+          // Use a post-frame callback to avoid calling setState() during the
+          // initial build phase. Right now this is being called in the wrong
+          // order. (Refreshing the browser with the initial route set to
+          // '/settings' invokes onGenerateRoute with '/' instead of
+          // '/settings'.)
+          SchedulerBinding.instance.addPostFrameCallback((dur) {
```

Comment:
We should set aside some time to talk about what you've needed to do here. I have no doubt that you've got the best available pattern, but having something this complicated in one of our samples is a DevRel smell, so to speak, and we should try to effect change to make the API better.

Reply:
Removed, since this is more of an experiment than a best-practice.